***Understanding Feature Distributions (Iris Dataset)
Context: Before training any ML model, we analyze feature distributions.***

In [2]:
import seaborn  as sns
import pandas as pd

_iris_data = sns.load_dataset("iris")
_iris_data

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa
...,...,...,...,...,...
145,6.7,3.0,5.2,2.3,virginica
146,6.3,2.5,5.0,1.9,virginica
147,6.5,3.0,5.2,2.0,virginica
148,6.2,3.4,5.4,2.3,virginica


Extract feature matrix as NumPy array

In [21]:
import numpy as np
iris_array_data = np.array(_iris_data)
iris_array_data

array([[5.1, 3.5, 1.4, 0.2, 'setosa'],
       [4.9, 3.0, 1.4, 0.2, 'setosa'],
       [4.7, 3.2, 1.3, 0.2, 'setosa'],
       [4.6, 3.1, 1.5, 0.2, 'setosa'],
       [5.0, 3.6, 1.4, 0.2, 'setosa'],
       [5.4, 3.9, 1.7, 0.4, 'setosa'],
       [4.6, 3.4, 1.4, 0.3, 'setosa'],
       [5.0, 3.4, 1.5, 0.2, 'setosa'],
       [4.4, 2.9, 1.4, 0.2, 'setosa'],
       [4.9, 3.1, 1.5, 0.1, 'setosa'],
       [5.4, 3.7, 1.5, 0.2, 'setosa'],
       [4.8, 3.4, 1.6, 0.2, 'setosa'],
       [4.8, 3.0, 1.4, 0.1, 'setosa'],
       [4.3, 3.0, 1.1, 0.1, 'setosa'],
       [5.8, 4.0, 1.2, 0.2, 'setosa'],
       [5.7, 4.4, 1.5, 0.4, 'setosa'],
       [5.4, 3.9, 1.3, 0.4, 'setosa'],
       [5.1, 3.5, 1.4, 0.3, 'setosa'],
       [5.7, 3.8, 1.7, 0.3, 'setosa'],
       [5.1, 3.8, 1.5, 0.3, 'setosa'],
       [5.4, 3.4, 1.7, 0.2, 'setosa'],
       [5.1, 3.7, 1.5, 0.4, 'setosa'],
       [4.6, 3.6, 1.0, 0.2, 'setosa'],
       [5.1, 3.3, 1.7, 0.5, 'setosa'],
       [4.8, 3.4, 1.9, 0.2, 'setosa'],
       [5.0, 3.0, 1.6, 0.

Compute: Mean Median Standard deviation Variance (for each feature)

In [3]:
numerical_iris_data = _iris_data.select_dtypes(include='number')
numerical_iris_describe = pd.DataFrame({
    'mean': numerical_iris_data.mean(),
    'median': numerical_iris_data.median(),
    'standard deviation': numerical_iris_data.std(),
    'Variance': numerical_iris_data.var()
})
numerical_iris_describe

,mean,median,standard deviation,Variance
sepal_length,5.843333,5.80,0.828066,0.685694
sepal_width,3.057333,3.00,0.435866,0.189979
petal_length,3.758000,4.35,1.765298,3.116278
petal_width,1.199333,1.30,0.762238,0.581006


Identify: Which feature has highest variability and why it matters in ML
here, Standard deviation compute variability.

**petal_length feature has highest variability (1.765298) compared to others, as it has higher length variance corresponding to the different species.**<br>
**variability** - ____used in data analysis shows which feature has more information and help to detect outliner___<br>

**higher variability** - ___value changes a lot across sample__<br>

___why it matters in ML?___<br>
**Model learn by mapping, if feature barly change the feature has little to learn
                              if feature is constant cannot predict future output
                              but feature varies, it help to predict output**

Convert any one feature into shape (n,1) and explain why ML models expect this format

In [34]:
sepal_length_1d_array =  numerical_iris_data[['sepal_length']]
sepal_length_1d_array
#sepal_length_1d_array.shape

,sepal_length
0,5.1
1,4.9
2,4.7
3,4.6
4,5.0
...,...
145,6.7
146,6.3
147,6.5
148,6.2


why ML models expect this format --

**ML models are based on linear algebra and expect a feature matirx. y = Xw+b <br>
X= feature matrix<br>
w- weight vector<br>
b- Intercept <br>
row = sample<br>
column = feature<br>
if we pass matrix with (1,) or (150,) its a 1D array and model get confused furthermore couldn't able to identify the diredtion. <br>
suppose if we have (150,1) it's 150 sample with 1 feature but with (150,)<br>
for 1D case 150 sample with 1 feature or 1 feature with 150 sample and end up in error as model can not define feature-weight relationship.**
